cryptography/src/bls12381/scheme.rs

```diff
@@ -98,3 +98,502 @@ impl Scheme for Bls12381 {
+            vector_sign_8(),
+            vector_sign_9(),
+        ];
+
+        for (index, test_case) in tests_cases.iter().enumerate() {
+            test_vector_sign(
+                index + 1,
+                test_case.0.clone(),
+                test_case.1.clone(),
+                test_case.2.clone(),
+            );
+        }
+    }
+
+    #[test]
+    fn sign_zero_private_key() {
+        let v = vector_sign_10();
+        let private_key = PrivateKey::from(v.0);
+        let signer = <Bls12381 as Scheme>::from(private_key);
+        assert!(signer.is_none())
+    }
+
+    #[test]
+    fn verify() {
+        let tests_cases = [
+            vector_verify_1(),
+            vector_verify_2(),
+            vector_verify_3(),
+            vector_verify_4(),
+            vector_verify_5(),
+            vector_verify_6(),
+            vector_verify_7(),
+            vector_verify_8(),
+            vector_verify_9(),
+            vector_verify_10(),
+            vector_verify_11(),
+            vector_verify_12(),
+            vector_verify_13(),
+            vector_verify_14(),
+            vector_verify_15(),
+            vector_verify_16(),
+            vector_verify_17(),
+            vector_verify_18(),
+            vector_verify_19(),
+            vector_verify_20(),
+            vector_verify_21(),
+            vector_verify_22(),
+            vector_verify_23(),
+            vector_verify_24(),
+            vector_verify_25(),
+            vector_verify_26(),
+            vector_verify_27(),
+            vector_verify_28(),
+            vector_verify_29(),
+        ];
+
+        for (index, test_case) in tests_cases.iter().enumerate() {
+            test_vector_verify(
+                index + 1,
+                test_case.0.clone(),
+                test_case.1.clone(),
+                test_case.2.clone(),
+                test_case.3,
+            );
+        }
+    }
+
+    /// Runs the provided `sign` vector test.
+    fn test_vector_sign(test_id: usize, private_key: Vec<u8>, message: Vec<u8>, output: Vec<u8>) {
+        let private_key = PrivateKey::from(private_key);
+        let mut signer =
+            <Bls12381 as Scheme>::from(private_key).expect("unable to deserialize private key");
+        let signature = signer.sign(None, &message);
+        assert_eq!(signature, output, "vector_sign_{} failed", test_id);
+    }
+
+    /// Runs the provided `verfify` vector test.
+    fn test_vector_verify(
+        test_id: usize,
+        public_key: Vec<u8>,
+        message: Vec<u8>,
+        signature: Vec<u8>,
+        success: bool,
+    ) {
+        let public_key = PublicKey::from(public_key);
+        let signature = Bytes::from(signature);
+        let res = Bls12381::verify(None, &message, &public_key, &signature);
+        assert_eq!(res, success, "vector_verify_{}", test_id);
+    }
+
+    /// Generates vector from hex encoded data.
+    fn generate_vector(s1: &str, s2: &str, s3: &str) -> (Vec<u8>, Vec<u8>, Vec<u8>) {
```

Comment:
nit: Let's call this `parse_vector`. Please update the args to be the name of what they are `private_key`, `public_key`, `signature` (I think?)

Reply:
You can also have this return `PrivateKey`, `PublicKey`, and `Signature` types (which just wrap `Bytes`).

Reply:
humm as we have moved the `bool` returned value outside of `vector_verify()`; `vector_sign()` and `vector_verify()` were doing the exact same logic, so it has been merged into `generate_vector()`. The only difference is the semantic of args passed to `generate_vector()`

for signing context it is: `private_key, message, signature`
for verifying context it is: `public_key, message, signature`

Reply:
To get the right arg names + response types, it's worth it IMO to return to using separate functions for parsing.

That's on me for suggesting that tweak without appreciating the downstream impact. Happy to make the change myself to get this wrapped up (don't mean to waste your time with duplicate work 😅). Just lmk!

Reply:
Don't worry, I change it ! 🙂